File: Quorum3Plugins/GamePlugin/JavaGame/src/plugins/quorum/Libraries/Game/Graphics/Mesh.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package plugins.quorum.Libraries.Game.Graphics;

import java.nio.FloatBuffer;
import java.nio.ShortBuffer;
import plugins.quorum.Libraries.Game.GameState;
import plugins.quorum.Libraries.Game.GameRuntimeError;
import plugins.quorum.Libraries.Game.libGDX.ShaderProgram;
import quorum.Libraries.Compute.Matrix4_;
import quorum.Libraries.Compute.Vector3;

import quorum.Libraries.Game.Graphics.Mesh_;
import quorum.Libraries.Game.BoundingBox_;
import quorum.Libraries.Game.Graphics.VertexAttribute_;

/**
 *
 * @author alleew
 */
public class Mesh 
{
    public java.lang.Object me_ = null;
    private quorum.Libraries.Game.Graphics.Mesh quorumMesh = null;
    
    // Static vector used during calculations.
    private static Vector3 calcVector = new Vector3();
    
    /*
    An action used by the Quorum mesh class during creation. This will ensure
    the plugin has a reference to its corresponding class.
    */
    public void SetQuorumReference(Mesh_ mesh)
    {
        quorumMesh = (quorum.Libraries.Game.Graphics.Mesh)mesh;
    }
    
    public void Render(ShaderProgram shader, int primitiveType) 
    {
        Render(shader, primitiveType, 0, quorumMesh.indices.GetMaxSize() > 0 ? quorumMesh.indices.GetSize() : quorumMesh.vertices.GetSize(), 
            quorumMesh.autoBind);
    }
    
    public void Render(ShaderProgram shader, int primitiveType, int offset, int count)
    {
        Render(shader, primitiveType, offset, count, quorumMesh.autoBind);
    }
    
    public void Render(ShaderProgram shader, int primitiveType, int offset, int count, boolean autoBind) 
    {
        if (count == 0) 
            return;

        if (autoBind) 
            Bind(shader);

        if (quorumMesh.isVertexArray) 
        {
            if (quorumMesh.indices.GetSize() > 0) 
            {
                ShortBuffer buffer = ((quorum.Libraries.Game.Graphics.IndexData)quorumMesh.indices).plugin_.GetBuffer();
                int oldPosition = buffer.position();
                int oldLimit = buffer.limit();
                buffer.position(offset);
                buffer.limit(offset + count);
                GameState.nativeGraphics.glDrawElements(primitiveType, count, GraphicsManager.GL_UNSIGNED_SHORT, buffer);
                buffer.position(oldPosition);
                buffer.limit(oldLimit);
            } 
            else
            {
                GameState.nativeGraphics.glDrawArrays(primitiveType, offset, count);
            }
        }
        else
        {
            if (quorumMesh.indices.GetSize() > 0)
            {
                GameState.nativeGraphics.glDrawElements(primitiveType, count, GraphicsManager.GL_UNSIGNED_SHORT, offset * 2);
            }
            else
            {
                GameState.nativeGraphics.glDrawArrays(primitiveType, offset, count);
            }
        }

        if (autoBind)
            Unbind(shader);
    }
    
    public void Bind(final ShaderProgram shader)
    {
        Bind(shader, null);
    }
    
    public void Bind(final ShaderProgram shader, final int[] locations)
    {
        ((quorum.Libraries.Game.Graphics.VertexBufferObject)quorumMesh.vertices).plugin_.Bind(shader, locations);
        if (quorumMesh.indices.GetSize() > 0)
            ((quorum.Libraries.Game.Graphics.IndexBufferObject)quorumMesh.indices).plugin_.Bind();
    }
    
    public void Unbind(final ShaderProgram shader)
    {
        Unbind(shader, null);
    }
    
    public void Unbind(final ShaderProgram shader, final int[] locations)
    {
        ((quorum.Libraries.Game.Graphics.VertexBufferObject)quorumMesh.vertices).plugin_.Unbind(shader, locations);
        if (quorumMesh.indices.GetSize() > 0)
            ((quorum.Libraries.Game.Graphics.IndexBufferObject)quorumMesh.indices).plugin_.Unbind();
    }
    
    public BoundingBox_ CalculateBoundingBox(BoundingBox_ box)
    {
        box.Invalidate();
        
        int numVertices = quorumMesh.GetVerticesCount();
        if (numVertices == 0)
            throw new GameRuntimeError("There were no vertices defined for this Mesh!");
        
        final FloatBuffer verts = ((quorum.Libraries.Game.Graphics.VertexBufferObject)quorumMesh.GetVertexData()).plugin_.GetBuffer();
        
        VertexAttribute_ posAttrib = quorumMesh.GetVertexAttributes().GetAttribute(quorumMesh.GetVertexAttributes().Get_Libraries_Game_Graphics_VertexAttributes__POSITION_());
        final int offset = posAttrib.Get_Libraries_Game_Graphics_VertexAttribute__offset_() / 4;
        final int vertexSize = quorumMesh.vertices.GetAttributes().Get_Libraries_Game_Graphics_VertexAttributes__vertexSize_();
        int index = offset;
        
        switch (posAttrib.Get_Libraries_Game_Graphics_VertexAttribute__componentCount_())
        {
            case 1:
                for (int i = 0; i < numVertices; i++)
                {
                    box.Extend(verts.get(index), 0, 0);
                    index += vertexSize;
                }
                break;
                
            case 2:
                for (int i = 0; i < numVertices; i++)
                {
                    box.Extend(verts.get(index), verts.get(index + 1), 0);
                    index += vertexSize;
                }
                break;
                
            case 3:
                for (int i = 0; i < numVertices; i++)
                {
                    box.Extend(verts.get(index), verts.get(index + 1), verts.get(index + 2));
                    index += vertexSize;
                }
                break;
        }
        
        return box;
    }
    
    public BoundingBox_ ExtendBoundingBox(BoundingBox_ box, int offset, int count, Matrix4_ transform)
    {
        int numIndices = quorumMesh.GetIndicesCount();
        
        if (offset < 0 || count < 1 || offset + count > numIndices)
            throw new GameRuntimeError("Invalid parameter(s) to ExtendBoundingBox - offset = " + offset + ", count = " + count + "max = " + numIndices);

        final FloatBuffer verts = ((quorum.Libraries.Game.Graphics.VertexBufferObject)quorumMesh.vertices).plugin_.GetBuffer();
        final ShortBuffer index = ((quorum.Libraries.Game.Graphics.IndexBufferObject)quorumMesh.indices).plugin_.GetBuffer();
        final VertexAttribute_ posAttrib = quorumMesh.GetVertexAttributes().GetAttribute(quorumMesh.GetVertexAttributes().Get_Libraries_Game_Graphics_VertexAttributes__POSITION_());
        
        final int posOffset = posAttrib.Get_Libraries_Game_Graphics_VertexAttribute__offset_() / 4;
        final int vertexSize = quorumMesh.vertices.GetAttributes().Get_Libraries_Game_Graphics_VertexAttributes__vertexSize_() / 4;
        final int end = offset + count;
        
        switch (posAttrib.Get_Libraries_Game_Graphics_VertexAttribute__componentCount_())
        {
            case 1:
                for (int i = offset; i < end; i++)
                {
                    final int idx = index.get(i) * vertexSize + posOffset;
                    calcVector.Set(verts.get(idx), 0, 0);
                    if (transform != null)
                        calcVector.Multiply(transform);
                    
                    box.Extend(calcVector);
                }
                break;
                
            case 2:
                for (int i = offset; i < end; i++)
                {
                    final int idx = index.get(i) * vertexSize + posOffset;
                    calcVector.Set(verts.get(idx), verts.get(idx + 1), 0);
                    if (transform != null)
                        calcVector.Multiply(transform);
                    box.Extend(calcVector);
                }
                break;
                
            case 3: 
                for (int i = offset; i < end; i++)
                {
                    final int idx = index.get(i) * vertexSize + posOffset;
                    calcVector.Set(verts.get(idx), verts.get(idx + 1), verts.get(idx + 2));
                    if (transform != null)
                        calcVector.Multiply(transform);
                    box.Extend(calcVector);
                }
            
                break;
        }
        
        return box;
    }
    
}

File: Quorum/Library/Standard/Plugins/javascript/Libraries/Game/Graphics/Mesh.js
/* global plugins_quorum_Libraries_Game_GameStateManager_ */

function plugins_quorum_Libraries_Game_Graphics_Mesh_() 
{
    this.quorumMesh = null;
    this.calcVector = new quorum_Libraries_Compute_Vector3_();
    
    this.SetQuorumReference$quorum_Libraries_Game_Graphics_Mesh = function(mesh) 
    {
        this.quorumMesh = mesh;
    };
    
    /* The shader parameter must be a valid ShaderProgram, and the primitiveType
     * parameter must be one of the following constant values accessible from
     * the gl property in WebGraphics:
     *      POINTS: Draws a single dot at each vertex.
     *      LINE_STRIP: Draws a straight line to the next vertex.
     *      LINE_LOOP: Draws a straight line to the next vertex, and connects the last vertex back to the first.
     *      LINES: Draws a line between a pair of vertices.
     *      TRIANGLE_STRIP: Draws a triangle between the first three vertices. Each successive vertex forms a triangle with the last two vertices used.
     *      TRIANGLE_FAN: Draws a triangle between the first three vertices. Each successive vertex forms a triangle with the first vertex and the last vertex used.
     *      TRIANGLES: Draws a triangle for a group of three vertices.
     *      
     * If the offset, count, or autoBind values are undefined, default values
     * will be provided using the contents of the Quorum Mesh object.
     * 
     * @param {ShaderProgram} shader
     * @param {integer} primitiveType
     * @param {integer} offset
     * @param {integer} count
     * @param {boolean} autoBind
     * @returns {undefined}
     */
    this.Render = function(shader, primitiveType, offset, count, autoBind)
    {
        if (offset === undefined)
            offset = 0;
        
        if (count === undefined)
        {
            if (this.quorumMesh.indices.GetSize() > 0)
                count = this.quorumMesh.indices.GetSize();
            else
                count = this.quorumMesh.vertices.GetSize();
        }
        
        if (count === 0)
            return;
        
        var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
        
        if (autoBind === undefined)
            autoBind = this.quorumMesh.autoBind;
        
        if (autoBind)
            this.Bind(shader);
        
        if (this.quorumMesh.indices.GetSize() > 0)
        {
            if (this.quorumMesh.indices.plugin_.supports32bits)
                graphics.glDrawElements(primitiveType, count, graphics.gl.UNSIGNED_INT, offset * 4);
            else
                graphics.glDrawElements(primitiveType, count, graphics.gl.UNSIGNED_SHORT, offset * 2);
        }
        else
        {
            graphics.glDrawArrays(primitiveType, offset, count);
        }
      
        if (autoBind)
            this.Unbind(shader);
    };
    
    this.Bind = function(shader, locations)
    {
        this.quorumMesh.vertices.plugin_.Bind(shader, locations);
        if (this.quorumMesh.indices.GetSize() > 0)
            this.quorumMesh.indices.plugin_.Bind();
    };
    
    this.Unbind = function(shader, locations)
    {
        this.quorumMesh.vertices.plugin_.Unbind(shader, locations);
        if (this.quorumMesh.indices.GetSize() > 0)
            this.quorumMesh.indices.plugin_.Unbind();
    };
    
    this.CalculateBoundingBox$quorum_Libraries_Game_BoundingBox = function(box) 
    {
        box.Invalidate();
        
        var numVertices = this.quorumMesh.GetVerticesCount();
        if (numVertices === 0)
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("There were no vertices defined for this Mesh!");
            throw exceptionInstance_;
        }
        
        var verts = this.quorumMesh.vertices.plugin_.GetBuffer();
        
        var posAttrib = this.quorumMesh.GetVertexAttributes().FindByUsage$quorum_integer(this.quorumMesh.GetVertexAttributes().POSITION);
        var offset = posAttrib.offset / 4;
        var vertexSize = this.quorumMesh.vertices.GetAttributes().vertexSize;
        var index = offset;
        
        switch (posAttrib.componentCount)
        {
            case 1:
                for (var i = 0; i < numVertices; i++)
                {
                    box.Extend$quorum_number$quorum_number$quorum_number(verts[index], 0, 0);
                    index = index + vertexSize;
                }
                break;
                
            case 2:
                for (var i = 0; i < numVertices; i++)
                {
                    box.Extend$quorum_number$quorum_number$quorum_number(verts[index], verts[index + 1], 0);
                    index = index + vertexSize;
                }
                break;
                
            case 3:
                for (var i = 0; i < numVertices; i++)
                {
                    box.Extend$quorum_number$quorum_number$quorum_number(verts[index], verts[index + 1], verts[index + 2]);
                    index = index + vertexSize;
                }
                break;
        }
        return box;
    };
    
    this.ExtendBoundingBox$quorum_Libraries_Game_BoundingBox$quorum_integer$quorum_integer$quorum_Libraries_Compute_Matrix4 = function(box, offset, count, transform) 
    {
        var numIndices = this.quorumMesh.GetIndicesCount();
        
        if (offset < 0 || count < 1 || offset + count > numIndices)
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Invalid parameter(s) to ExtendBoundingBox - offset = " + offset + ", count = " + count + ", max = " + numIndices);
            throw exceptionInstance_;
        }
        
        var vertices = this.quorumMesh.vertices.plugin_.GetBuffer();
        var indices = this.quorumMesh.indices.plugin_.GetBuffer();
        
        var posAttrib = this.quorumMesh.GetVertexAttributes().FindByUsage$quorum_integer(this.quorumMesh.GetVertexAttributes().POSITION);
        
        var posOffset = posAttrib.offset / 4;
        var vertexSize = this.quorumMesh.vertices.GetAttributes().vertexSize / 4;
        var end = offset + count;
        
        switch (posAttrib.componentCount)
        {
            case 1:
                for (var i = offset; i < end; i++)
                {
                    var idx = indices[i] * vertexSize + posOffset;
                    this.calcVector.Set$quorum_number$quorum_number$quorum_number(vertices[idx], 0, 0);
                    if (transform !== null && transform !== undefined)
                        this.calcVector.Multiply$quorum_Libraries_Compute_Matrix4(transform);
                    
                    box.Extend$quorum_Libraries_Compute_Vector3(this.calcVector);
                }
                break;
                
            case 2:
                for (var i = offset; i < end; i++)
                {
                    var idx = indices[i] * vertexSize + posOffset;
                    this.calcVector.Set$quorum_number$quorum_number$quorum_number(vertices[idx], vertices[idx + 1], 0);
                    if (transform !== null && transform !== undefined)
                        this.calcVector.Multiply$quorum_Libraries_Compute_Matrix4(transform);
                    
                    box.Extend$quorum_Libraries_Compute_Vector3(this.calcVector);
                }
                break;
                
            case 3:
                for (var i = offset; i < end; i++)
                {
                    var idx = indices[i] * vertexSize + posOffset;
                    this.calcVector.Set$quorum_number$quorum_number$quorum_number(vertices[idx], vertices[idx + 1], vertices[idx + 2]);
                    if (transform !== null && transform !== undefined)
                        this.calcVector.Multiply$quorum_Libraries_Compute_Matrix4(transform);
                    
                    box.Extend$quorum_Libraries_Compute_Vector3(this.calcVector);
                }
                break;
        }
    };
    
    this.AddReloadableMesh = function()
    {
        // Do nothing, this isn't necessary on the web.
    };
    
    this.RemoveReloadableMesh = function()
    {
        // Do nothing, this isn't necessary on the web.
    };
}
